test/e2e/etcd_multi_node_test.go

```diff
@@ -0,0 +1,441 @@
+//
+// checkEventuallyEtcdRollingUpdateDone ensures rolling updates of etcd resources(etcd/sts) is done.
+func checkEventuallyEtcdRollingUpdateDone(ctx context.Context, cl client.Client, logger logr.Logger, etcd *v1alpha1.Etcd,
+	oldStsObservedGeneration, oldEtcdObservedGeneration int64) {
+	Eventually(func() error {
+		sts := &appsv1.StatefulSet{}
+		ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), sts)).To(Succeed())
+
+		if sts.Status.ObservedGeneration <= oldStsObservedGeneration {
+			return fmt.Errorf("waiting for statefulset rolling update to complete %d pods at revision %s",
+				sts.Status.UpdatedReplicas, sts.Status.UpdateRevision)
+		}
+
+		if sts.Status.UpdatedReplicas != *sts.Spec.Replicas {
+			return fmt.Errorf("waiting for statefulset rolling update to complete, UpdatedReplicas is %d, but expected to be %d",
+				sts.Status.UpdatedReplicas, *sts.Spec.Replicas)
+		}
+
+		ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+		if *etcd.Status.ObservedGeneration <= oldEtcdObservedGeneration {
+			return fmt.Errorf("waiting for etcd %q rolling update to complete", etcd.Name)
+		}
+		return nil
+	}, timeout*3, pollingInterval).Should(BeNil())
+}
+
+// hibernateAndCheckEtcd scales down etcd replicas to zero and ensures
+func hibernateAndCheckEtcd(ctx context.Context, cl client.Client, logger logr.Logger, etcd *v1alpha1.Etcd) {
+	ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+	etcd.Spec.Replicas = 0
+	etcd.SetAnnotations(
+		map[string]string{
+			v1beta1constants.GardenerOperation: v1beta1constants.GardenerOperationReconcile,
+		})
+	ExpectWithOffset(1, cl.Update(ctx, etcd)).ShouldNot(HaveOccurred())
+	logger.Info("Waiting to hibernate")
+
+	logger.Info("Checking statefulset")
+	Eventually(func() error {
+		sts := &appsv1.StatefulSet{}
+		ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), sts)).To(Succeed())
+
+		if sts.Status.ReadyReplicas != 0 {
+			return fmt.Errorf("sts %s not ready", etcd.Name)
+		}
+		return nil
+	}, timeout*3, pollingInterval).Should(BeNil())
+
+	logger.Info("Checking etcd")
+	Eventually(func() error {
+		etcd := getEmptyEtcd(etcd.Name, namespace)
+		err := cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)
+		if err != nil {
+			return err
+		}
+
+		if etcd != nil && etcd.Status.Ready != nil {
+			if *etcd.Status.Ready != true {
+				return fmt.Errorf("etcd %s is not ready", etcd.Name)
+			}
+		}
+
+		if etcd.Status.ClusterSize == nil {
+			return fmt.Errorf("etcd %s cluster size is empty", etcd.Name)
+		}
+		// TODO: uncomment me once scale down is supported,
+		// currently ClusterSize is not updated while scaling down.
+		// if *etcd.Status.ClusterSize != 0 {
+		// 	return fmt.Errorf("etcd %q cluster size is %d, but expected to be 0",
+		// 		etcdName, *etcd.Status.ClusterSize)
+		// }
+
+		for _, c := range etcd.Status.Conditions {
+			if c.Status != v1alpha1.ConditionUnknown {
```

Comment:
FYI: This will change with https://github.com/gardener/etcd-druid/pull/415, i.e. `status` can also remain `true`.